### PR TITLE
Behavior manager class

### PR DESCRIPTION
**Pull request by ethanmusser**
An abstract class for implementing `RobotBase`-derived robot behaviors.

- Provides a `BehaviorManager` class in the `kodlab` namespace to maintain one or several `Behavior`-derived behaviors and manage switching between them.
  - `BehaviorManager` is templated for a `RobotBase`-derived robot, as are the behaviors it holds.
  - Stores behaviors internally in a variable-length container and maintains a default behavior at index 0 of this container (by default, this is the `kodlab::OffBehavior`).
  - Provides methods for adding behaviors and switching between them, including waiting to switch until the running behavior is ready.

Note that this PR is 2/3 in the behavior implementation series and depends on the `Behavior` class introduced in #39.

Features to be implemented in future work include:

- Event-based behavior switching.
- Maintaining a timeout behavior that, if the running behavior does not permit switching in within a set period of time, will be switched to.

### CODE REVIEW

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:68]: Please add a map from name to index

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:105]: provide an overload that uses the map

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [reply include/kodlab_mjbots_sdk/behavior_manager.h:68]: (unordered map is probably best)

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:127]: I feel like we should not just warn when this happens.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:112]: nit: brackets

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:140]: I don't think init behavior is the correct function name. Maybe add behavior?

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:148]: What is the purpose of this function? When do you expect it to be used?

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [reply include/kodlab_mjbots_sdk/behavior_manager.h:140]: Proposal change to CreateBehavior and make it private.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [reply include/kodlab_mjbots_sdk/behavior_manager.h:148]: see above comment

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:259]: should specify index in function name. This makes me think it returns a pointer to the behavior.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:287]: `get_behavior_names.

Current function name makes me think its the vector of behaviors.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:22]: lets think about making a name space and possibly a separate folder for behavior related code. It will make it easier to share the code you wrote when you apply for jobs.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:206]: `EraseBehaviors` Reset makes me think the it just sets the behaviors back to their default state rather than removing all of them.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:159]: Lets return the index for the behavior.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:105]: How do people feel about providing the type of behavior and the checking to make sure the relevant behavior is the correct type? IDK how trivial it would be to code, but I just want to minimize the number of times the code enters the wrong behavior.

**Review by ShaneRozenLevy — COMMENTED** [review]
Code looks to be off to a good start. A few comments about function names, and I think adding an unorded map from string to index would make it much easier to use.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:127]: For behaviors, I think that warning is sufficient.  The way that @jdcaporale and I use this involves a simple CLI which prompts for a mode number and publishes it.  It is not uncommon for me to incorrectly enter a mode that is out of range.  In this case, I would prefer just to be warned than have the robot shut off or send the motors into braking.

If you feel differently, perhaps we could add an optional parameter when constructing the `BehaviorManager` to dictate what the robot should do in this circumstance (i.e., warn, return to default behavior, or kill).

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:22]: Sounds good to me!  If it's alright with you, I will wait and do this in a separate PR once #40 and #41 are merged since we have already merged #39 (`Behavior` and `IOBehavior`) in the current `kodlab` namespace.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:105]: This would definitely be safer, but it would require additional, redundant changes to user-defined `ProcessInput` methods to implement in full (e.g., defining a switch with cases for each expected behavior index, then running the appropriate `SetBehavior` template method.

Something that was talked about at some point was implementing a boolean behavior transition matrix, which the user could provide to dictate which inter-behavior transitions are acceptable.  I think this would not be particularly difficult to implement if it was set after all behaviors were added to the manager.  Adding as you go would be a little more tricky.  Either way, I would probably keep this as an optional feature.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:148]: This function constructs the behavior from the given arguments and calls the behavior's initialization method.  If the initialization method succeeds, the behavior is returned to be added to the behaviors list.  Otherwise, a `nullptr` is returned.

The name change and move to private sound good!

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:140]: Sounds good!

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [reply include/kodlab_mjbots_sdk/behavior_manager.h:105]: Ok, happy to leave the check as is. I think the string based argument will make things much easier to use.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:83]: Provided in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Added in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Fixed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Changed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Changed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Fixed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Renamed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Renamed in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Return value added in d5a6010ee947e62b6bbd122659ed3fcc02ff3224

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: An issue has been opened and assigned (#57).

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Agreed, definitely a good idea!  I will see about the transition matrix in a future issue.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:199]: hmm. Thats probably bad since -1 is the kill behavior index. lets make it an optional int.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:165]: lets use std::optional here

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:307]: std::optional. Other option would be to return the iterator, and have the use check if its equal to the end, but the std::optional is a clearer design

**Review by ShaneRozenLevy — COMMENTED** [review]
Looking close to good. A few small changes, then I'll look everything over at should be good.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Haha probably for the best, fixed in 6bc58c22c7bfe4d156e3962308c3fbf73fc639c6

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Agreed `std::optional` is the way.  Added in 6bc58c22c7bfe4d156e3962308c3fbf73fc639c6

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:166]: Changed in 6bc58c22c7bfe4d156e3962308c3fbf73fc639c6

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:57]: https://www.wired.com/2015/11/null/

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:115]: In the current paradigm, is init ever called not right after the constructor?

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ShaneRozenLevy) [inline include/kodlab_mjbots_sdk/behavior_manager.h:140]: Don't pass int by reference: http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional

**Review by ShaneRozenLevy — APPROVED** [review]
LGTM only blocker is passing int by reference.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:167]: This is worse as far as formatting than the previous IMO.

Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:167]:
```suggestion
      if (idx) {
        SetBehavior(idx.value());
      }
```

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:118]:
```suggestion
    } else {
      return nullptr;
    }
```

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:221]: shouldnt these use `DEFAULT_IDX`?

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:225]: If the behavior fails to construct, would it have a get_name member?

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (jdcaporale) [inline include/kodlab_mjbots_sdk/behavior_manager.h:84]: I'm torn. My gut says names_ that we shouldn't keep two vectors that match index the way names_ and behaviors_ do. 
Instead just query behaviors_.get_name().

It is nice for convenience, but a possible vector for errors and bugs. 
Most of the places that use names switch effortlessly. 
The member function get_behavior_names() will need a for loop or transform or something. 

The map is a useful thing to go the other way though so happy have that even though it also has a similar issue.

**Review by jdcaporale — APPROVED** [review]
Couple possible bugs and some philosophical and formatting changes, but all small so LGTM after you deal with that.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:115]: No, currently `Init` is only called right after the constructor.  `Init` called outside the constructor as the [style guide](https://drake.mit.edu/styleguide/cppguide.html#Doing_Work_in_Constructors) discourages calling virtual methods in constructors (for good reason).  This separation is also maintained in case a `BehaviorManager` child wishes to call `Init` for all contained behaviors at the same time.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:225]: A `false` return value from `ConstructBehavior` actually indicates that the behavior failed to initialize (i.e., the `Init` method returned `false`).  If the behavior fails to construct due to something in runtime, the program will terminate with a segmentation fault.  However, all other constructor issues should be caught at compile time.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h:84]: This was added to avoid repeated calls to `Behavior::get_name`, particularly the looping in `get_behavior_named`.  I could go either way.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Fixed in e1e7ee036a97fdee4c755d57904d3f9dc82879e4

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: changed in 57a65848a1b429d69bd735eddc489b60878e210b

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Clarified in 57a65848a1b429d69bd735eddc489b60878e210b

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Changed to `el` (element) in 57a65848a1b429d69bd735eddc489b60878e210b.

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Formatted per style guide in 57a65848a1b429d69bd735eddc489b60878e210b

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Formatted per style guide in 57a65848a1b429d69bd735eddc489b60878e210b

File: include/kodlab_mjbots_sdk/behavior_manager.h
Comment (ethanmusser) [reply include/kodlab_mjbots_sdk/behavior_manager.h]: Yes they should, fixed in 57a65848a1b429d69bd735eddc489b60878e210b

**Pull request merged**
